sae/blocks.go

```diff
@@ -5,42 +5,240 @@ package sae
+// [VM.SetPreference] as the parent. The block context MAY be nil.
+func (vm *VM) BuildBlock(ctx context.Context, bCtx *block.Context) (*blocks.Block, error) {
+	parent := vm.preference.Load()
+
+	baseFee := uint256.NewInt(0) // TODO(arr4n) source this from the parent's worst-case *end* of execution
+	filter := txpool.PendingFilter{
+		BaseFee: baseFee,
+	}
+	return vm.buildBlock(ctx, bCtx, parent, vm.now(), vm.mempool.TransactionsByPriority(filter))
+}
+
+var errExecutionLagging = errors.New("execution lagging for settlement")
+
+// buildBlock implements the block-building logic shared by [VM.BuildBlock] and
+// [VM.VerifyBlock].
+func (vm *VM) buildBlock(ctx context.Context, bCtx *block.Context, parent *blocks.Block, blockTime time.Time, candidates []*txgossip.LazyTransaction) (*blocks.Block, error) {
+	log := vm.log().With(
+		zap.Uint64("parent_height", parent.Height()),
+		zap.Stringer("parent_hash", parent.Hash()),
+		zap.Time("block_time", blockTime),
+	)
+
+	settleAt := unix(blockTime.Add(-params.Tau))
+	lastSettled, ok, err := blocks.LastToSettleAt(settleAt, parent)
+	if err != nil {
+		return nil, err
+	}
+	if !ok {
+		log.Warn("Execution lagging when determining last block to settle")
+		return nil, errExecutionLagging
+	}
+
+	// Although, at the time of writing this, the implementation of
+	// [blocks.LastToSettleAt] guarantees that the returned block has finished
+	// execution, this isn't a stated invariant and it's possible for it to be
+	// more aggressive in the future.
+	wCtx, cancel := context.WithTimeout(ctx, 10*time.Millisecond)
+	defer cancel()
+	if err := lastSettled.WaitUntilExecuted(wCtx); err != nil {
+		log.Warn(
+			"Execution lagging for settlement artefacts",
+			zap.Uint64("settling_height", lastSettled.Height()),
+		)
+		return nil, fmt.Errorf("%w: settling block %d: %T.WaitUntilExecuted(): %v", errExecutionLagging, lastSettled.Height(), lastSettled, err)
+	}
+
+	sdb, err := state.New(lastSettled.PostExecutionStateRoot(), vm.exec.StateCache(), nil)
+	if err != nil {
+		b := lastSettled
+		return nil, fmt.Errorf("state.New([post-execution state root %#x of block %d], ...): %v", b.PostExecutionStateRoot(), b.Height(), err)
+	}
+
+	// TODO(arr4n) worst-case tx validation will happen here. This is just a
+	// sketch of the process.
+	history := parent.WhenChildSettles(lastSettled)
+	var receipts types.Receipts
+	for _, b := range history {
+		receipts = append(receipts, b.Receipts()...)
+	}
+
+	clock := lastSettled.ExecutedByGasTime()
+	for _, b := range history {
+		clock.BeforeBlock(vm.hooks, b.Header())
+
+		var lim gas.Gas
+		for _, tx := range b.Transactions() {
+			lim += gas.Gas(tx.Gas())
+		}
+		if err := clock.AfterBlock(lim, vm.hooks, b.Header()); err != nil {
+			return nil, err
+		}
+	}
+
+	hdr := &types.Header{
+		ParentHash: parent.Hash(),
+		Number:     new(big.Int).Add(parent.Number(), common.Big1),
+		Time:       unix(blockTime),
+		Root:       lastSettled.PostExecutionStateRoot(),
+	}
```

Comment:
Probably not an issue for this PR, but this doesn't expose sufficient functionality to do things like populating the ms timestamps, or updating the gas target, ect.